coderd/workspaceusage/tracker.go

```diff
@@ -0,0 +1,221 @@
+	"cdr.dev/slog"
+	"cdr.dev/slog/sloggers/sloghuman"
+)
+
+var DefaultFlushInterval = 60 * time.Second
+
+// Store is a subset of database.Store
+type Store interface {
+	BatchUpdateWorkspaceLastUsedAt(context.Context, database.BatchUpdateWorkspaceLastUsedAtParams) error
+}
+
+// Tracker tracks and de-bounces updates to workspace usage activity.
+// It keeps an internal map of workspace IDs that have been used and
+// periodically flushes this to its configured Store.
+type Tracker struct {
+	log       slog.Logger      // you know, for logs
+	flushLock sync.Mutex       // protects m
+	m         *uuidSet         // stores workspace ids
+	s         Store            // for flushing data
+	tickCh    <-chan time.Time // controls flush interval
+	stopTick  func()           // stops flushing
+	stopCh    chan struct{}    // signals us to stop
+	stopOnce  sync.Once        // because you only stop once
+	doneCh    chan struct{}    // signifies that we have stopped
+	flushCh   chan int         // used for testing.
+}
+
+// New returns a new Tracker. It is the caller's responsibility
+// to call Close().
+func New(s Store, opts ...Option) *Tracker {
+	hb := &Tracker{
+		log:      slog.Make(sloghuman.Sink(os.Stderr)),
+		m:        &uuidSet{},
+		s:        s,
+		tickCh:   nil,
+		stopTick: nil,
+		stopCh:   make(chan struct{}),
+		doneCh:   make(chan struct{}),
+		flushCh:  nil,
+	}
+	for _, opt := range opts {
+		opt(hb)
+	}
+	if hb.tickCh == nil && hb.stopTick == nil {
+		ticker := time.NewTicker(DefaultFlushInterval)
+		hb.tickCh = ticker.C
+		hb.stopTick = ticker.Stop
+	}
+	return hb
+}
+
+type Option func(*Tracker)
+
+// WithLogger sets the logger to be used by Tracker.
+func WithLogger(log slog.Logger) Option {
+	return func(h *Tracker) {
+		h.log = log
+	}
+}
+
+// WithFlushInterval allows configuring the flush interval of Tracker.
+func WithFlushInterval(d time.Duration) Option {
+	return func(h *Tracker) {
+		ticker := time.NewTicker(d)
+		h.tickCh = ticker.C
+		h.stopTick = ticker.Stop
+	}
+}
+
+// WithFlushChannel allows passing a channel that receives
+// the number of marked workspaces every time Tracker flushes.
+// For testing only and will panic if used outside of tests.
+func WithFlushChannel(c chan int) Option {
+	if flag.Lookup("test.v") == nil {
+		panic("developer error: WithFlushChannel is not to be used outside of tests.")
+	}
+	return func(h *Tracker) {
+		h.flushCh = c
+	}
+}
+
+// WithTickChannel allows passing a channel to replace a ticker.
+// For testing only and will panic if used outside of tests.
+func WithTickChannel(c chan time.Time) Option {
+	if flag.Lookup("test.v") == nil {
+		panic("developer error: WithTickChannel is not to be used outside of tests.")
+	}
+	return func(h *Tracker) {
+		h.tickCh = c
+		h.stopTick = func() {}
+	}
+}
+
+// Add marks the workspace with the given ID as having been used recently.
+// Tracker will periodically flush this to its configured Store.
+func (wut *Tracker) Add(workspaceID uuid.UUID) {
```

Comment:
We have the opportunity here to be precise by capturing the actual `last_used_at` (or, at least, the time the API was called).
What is the reason for dropping the precision?

I can imagine a scenario where a precision would be useful if not imperative for auditing.

Reply:
I'm piggy-backing an existing query `BatchUpdateWorkspaceLastUsedAt` (https://github.com/coder/coder/blob/main/coderd/database/queries/workspaces.sql#L430-L436). I was hoping to avoid the need for changes to the database layer here, and I'm honestly not sure how this query could be rewritten to batch update the workspaces table with a different `last_used_at` for every `id`.

Reply:
Fair enough 👍